File: labs/dm/2/I.java
import java.util.Scanner;

/**
 * @author Saveliy Bakturin
 *
 * Don't write off, if you don't wanna be banned!
 */

public class I {
    public static void main(final String... args) {
        final Scanner in = new Scanner(System.in);
        final int start = in.nextInt();
        in.nextLine();
        final StringBuilder sb = new StringBuilder();
        final byte[] matrix = new byte[200000];
        if (start == 1) {
            code(in, sb, matrix);
        } else {
            decode(in, sb, matrix);
        }
        System.out.println(sb);
    }

    private static int shift(final int in) {
        return in << 1;
    }

    private static void decode(final Scanner in, final StringBuilder sb, final byte[] matrix) {
        int error = 0;
        int temp = 1, uArray = 1, uString = 0;
        final String s = in.nextLine();
        in.close();
        do {
            if (uArray == temp) {
                temp *= 2;
            }
            matrix[uArray] = Byte.parseByte(String.valueOf(s.charAt(uString)));
            uArray++;
            uString++;
        } while (uString != s.length());
        int lastPosition;
        if (uArray > temp) {
            lastPosition = temp;
        } else {
            lastPosition = temp / 2;
        }
        temp = 1;
        do {
            int sum = 0;
            for (int i = temp + 1; i < uArray; i++) {
                if ((i & temp) == temp) {
                    sum ^= matrix[i];
                }
            }
            if (sum != matrix[temp]) {
                error += temp;
            }
            temp *= 2;
        } while (temp <= lastPosition);
        if (matrix[error] == 0) {
            matrix[error] = 1;
        } else {
            matrix[error] = 0;
        }
        temp = 1;
        for (int i = 1; i < uArray; i++) {
            if (temp == i) {
                temp *= 2;
            } else {
                sb.append(matrix[i]);
            }
        }
    }

    private static void code(final Scanner in, final StringBuilder sb, final byte[] matrix) {
        int temp = 1, uArray = 1, uString = 0;
        final String s = in.nextLine();
        in.close();
        do {
            if (uArray != temp) {
                matrix[uArray] = Byte.parseByte(String.valueOf(s.charAt(uString)));
                uArray++;
                uString++;
            } else {
                temp = shift(temp);
                uArray++;
            }
        } while (uString != s.length());
        int lastPosition;
        if (uArray > temp) {
            lastPosition = temp;
        } else {
            lastPosition = temp >> 1;
        }
        temp = 1;
        do {
            int sum = 0;
            for (int i = temp; i < uArray; i++) {
                if ((i & temp) == temp) {
                    sum ^= matrix[i];
                }
            }
            matrix[temp] = (byte) sum;
            temp = shift(temp);
        } while (temp <= lastPosition);
        for (int i = 1; i < uArray; i++) {
            sb.append(matrix[i]);
        }
    }
}

File: algorithms/graphs/I.cpp
#include <cstdio>
#include <set>
#include <cmath>
#include <vector>
#include <limits>

/**
 * @author Saveliy Bakturin
 * <p>
 * Don't write off, if you don't wanna be banned!
 */

double squared_distance(int const& x0, int const& y0, int const& x1, int const& y1) {
	return std::pow(x1 - x0, 2) + std::pow(y1 - y0, 2);
}

int main() {
	std::size_t n;
	std::scanf("%zu", &n);
	std::vector<int> x(n + 1), y(n + 1);
	std::vector<double> dist(n + 1);
	std::vector<bool> visited(n + 1);
	double sum = 0;
	auto get_min = [](std::vector<bool> &visited, std::vector<double> &dist, std::size_t const& n){
		double min = std::numeric_limits<double>::infinity();
		std::size_t index = 0;
		for (std::size_t i = 1; i < n + 1; i++) {
			if (!visited[i] && (dist[i] < min)) {
				min = dist[i];
				index = i;
			}
		}
		return index;
	};
	for (std::size_t i = 1; i < n + 1; i++) {
		std::scanf("%i %i", &x[i], &y[i]);
		dist[i] = std::numeric_limits<double>::infinity();
	}
	dist[1] = 0;
	for (std::size_t i = 1; i < n + 1; i++) {
		std::size_t v = get_min(visited, dist, n);
		sum += std::sqrt(dist[v]);
		for (std::size_t u = 1; u < n + 1; u++) {
			if (!visited[u] && v != u) dist[u] = std::min(dist[u], squared_distance(x[v], y[v], x[u], y[u]));
		}
		visited[v] = true;
	}
	std::printf("%.10lf", sum);
}
